sycl/doc/CompilerAndRuntimeDesign.md

```diff
@@ -787,18 +787,61 @@ The SPIR-V specific functions are implemented in for the SYCL host device here:
 
 ### Address spaces handling
 
-SYCL 1.2.1 language defines several address spaces where data can reside, the
-same as OpenCL - global, local, private and constant. From the spec: "In OpenCL
-C, these address spaces are manually specified using OpenCL-specific keywords.
-In SYCL, the device compiler is expected to auto-deduce the address space for
-pointers in common situations of pointer usage. However, there are situations
-where auto-deduction is not possible".
+SYCL-1.2.1 specification uses C++ templates to represent pointers to disjoint
+memory regions on an accelerator to enable compilation with standard C++
+toolchain and SYCL compiler toolchain.
+
+For instance:
+
+``` C++
+// check that SYCL mode is ON and we can use non-standard annotations
+#if defined(__SYCL_DEVICE_ONLY__)
+// GPU/accelerator implementation
+template <typename T, address_space AS> class multi_ptr {
+  // GetAnnotatedPointer<T, global>::type == "__attribute__((opencl_global)) T"
+  using pointer_t = typename GetAnnotatedPointer<T, AS>::type *;
```

Comment:
What is `GetAnnotatedPointer`, where is it defined? Maybe give a link?

Reply:
I'm trying to keep the code example short and I think that particular definition of `GetAnnotatedPointer` is not that important as how it maps it's parameters T and AS to `GetAnnotatedPointer<T, AS>::type` typename. I provided an example of such mapping in the comment above. T is arbitrary type and other AS values and applied similar to `opencl_global` showed in the example.
Do you think it's enough?

Reply:
so the `GetAnnotatedPointer` is an implementation detail, right? Why show it to the user at all then?

Reply:
This document describes DPC++ internal design details and it's intended for DPC++ developers rather than users, so it should be okay to put an implementation details here. This part was requested during code review in llorg, so that reviewers can better understand how OpenCL address space attributes are used in SYCL context.
I've added a reference to the source code with helper template class (`PtrValueType`) implementation.

Reply:
Maybe I am missing something big, but the code snippet at lines 801-803 doesn't clarify anything to me. Maybe you should add more text about what you are trying to convey.

Reply:
The intent of this code snippet is to demonstrate how OpenCL address space attributes are supposed to be used by SYCL implementation. I tried to come-up with minimal code snippet to make it easier to read than `multi_ptr` implementation in DPC++.
My thought was that `GetAnnotatedPointer` implementation is not relevant to this intent because the code snippet uses only `GetAnnotatedPointer<T, AS>::type` type. So the important question is "what type is GetAnnotatedPointer<T, AS>::type" and I tried to address this question in the comment. `GetAnnotatedPointer` is a helper class which provides a type alias annotated with OpenCL address space attribute. It's introduced to not duplicate the whole `multi_ptr` class. I hope this clarifies why the example is written this way.

To make it more clear I did following:
1. Created #3173 to give a better name for similar class in DPC++ implementation - `DecoratedType`
2. Used the same name in the example.
3. I also aligned `multi_ptr` class members with DPC++ implementation to make it easier to correlate this example with the implementation, although I avoided this correlation originally to have flexibility in using member names better suited for this example.

Now you should able find complete implementation of this helper class in DPC++ header if needed.

Alternative way to simplify the example is to get rid of `DecoratedType` helper class and provide an example of `multi_ptr` class specialization for one of the `address_space` values, which uses OpenCL address space attribute directly w/o a wrapper class. What do you think?

Reply:
OK this is more clear to me now (not so sure if it is also clear for a new reader).
The `DecoratedType` is only used in the code snippet, but not anywhere else.
Maybe change this brief source code comment to explicitly say that the `DecoratedType` is what the OpenCL-style attribute translates into internally:

// DecoratedType<T, global>::type == "__attribute__((opencl_global)) T"

Reply:
I'll add following comment:

```C++
// DecoratedType applies corresponding address space attribute to the type T
```